Battleship tutorial (#820)
midnightntwrk/midnight-docs: docs/tutorials/bship/smart-contract.mdx

---
SPDX-License-Identifier: Apache-2.0
copyright: This file is part of midnight-docs. Copyright (C) Midnight Foundation. Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0 Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
sidebar_label: "Battleship contract"

> title: "Battleship Contract"

Reviewer: ```suggestion
title: "Battleship contract"
```

description: "Learn about best practices while building a Battleship contract."
sidebar_position: 10
toc_max_heading_level: 2
---

> # Battleship Contract

Reviewer: ```suggestion
# Battleship contract
```

The Battleship tutorial is an intermediate level demonstration of the following features:
- Compact contracts as state machines
- Explicit state management
- Private state data (setting, getting, updating, verifying, maliciously manipulating)
- Operations on a `List`
- Intermediate Witness functionality
- Frontend tests with MidnightJS

It has two main components:
1. Compact contract
1. Test script

> This tutorial demonstrates writing safe and secure Compact contracts with a mix of public and private data while verifying the operation of the contract with MidnightJS in a local devnet test script. It is best consumed by writing each code block by hand rather than copy and pasting code.

Reviewer: ```suggestion
This tutorial demonstrates writing safe and secure Compact contracts with a mix of public and private data while verifying the operation of the contract with MidnightJS in a local devnet test script. It is best consumed by writing each code block by hand rather than copying and pasting code.
```

## Prerequisites

Before you begin this tutorial, ensure you have:
- [Installed the toolchain](../../getting-started/installation)
- Node.js v22+
- Completed a beginner tutorial

## Problem analysis

Battleship is a guessing game for two players. It is played on two individual grids with each player's ships marked on their grid. The location of a player's ships are hidden from the other player. Players alternate turns calling "shots" at the other player's ships with the objective of "sinking" those ships. The first player to correctly hit all of the opposing ships wins.  

Battleship is centered around the idea of private data (ships) and public data (shots, hits) being used in combination. The board game version comes with significant trust assumptions that make it vulnerable to malicious actors. A player could simply claim a MISS when it was in fact a HIT. They could move their ship in the middle of the game (change private state), or even look around the board at the location of the opposing player's ships to ensure a victory.

> Using blockchain, we can enforce the rules of this game easily by moving those trust assumptions on-chain. With Midnight, we can hide the private data in plain sight while verifying its validity in a players private state.

Reviewer: Do not use "we" or "us" in documentation. Apply throughout.

## Program design

> For the purposes of simplifying the code associated with this tutorial, we'll implement a more basic version of the game. The main differences in our version will be:

Reviewer: Do not use "we" or "us" in documentation. Apply throughout.

- Using a single number line instead of a grid
- A single number on the line represents the position of a ship
- Reduce the number of HITS required to win to two

### Operational steps

First, consider the operational components of the contract. What does the contract need to do? 

In order, it needs to:
1. Deploy the contract and allow player 1 to set their board
2. Allow player 2 to join the game and set their board
3. Player 1 shoots at board 2
4. Player 2 checks their board for a HIT or MISS
5. Player 2 shoots at board 1
6. Player 1 checks their board for HIT or MISS
7. Steps 3-6 continue until `board1Hits == 2 || board2Hits == 2`
8. Assign a winner

### Data: public vs private

One of the more challenging ideas in developing Midnight DApps is the mix of private and public data. Adhering to best-practice strategies is recommended to ensure that data is handled as intended and only made public when necessary. In this game, both players have the same needs in reference to their data:

> | Pulic Data | Private Data |

Reviewer 2: Pulic => Public

Also, convert the table heading to sentence case.

| ---------- | ------------ |

> | Dapp ID    | Address      |

Reviewer 2: Dapp => DApp

Review and apply throughout.

- Double shots
- Repeating a previous HIT to increment `hitCount`
- Claiming a MISS when it was in fact a HIT
- Changing ship locations mid-game

Smart contract developers *must be diligent in writing defensive programs.* The public and permissionless nature of blockchain means that contracts are accessible to anyone with a node command line. Intentionally guarding circuits through access control, input validation and explicit state management will create more secure programs than those without these considerations.

### State machines

Compact contracts are best thought of as state machines, so the next consideration is for the states of the contract. Custom states are needed for the following:
- BoardState
- ShotState
- WinState
- TurnState

By managing each of these states explicitly and pairing them with `assert` statements, access can be restricted to only the specific function needing to be available at any given state. For example, using `TurnState`, the contract can ensure that a player shoots only when it is their turn.

Turn order:  
`PLAYER_1_SHOOT` → `PLAYER_2_CHECK` → `PLAYER_2_SHOOT` → `PLAYER_1_CHECK` → repeat

> The `PLAYER_1_SHOOT` circuit will be available first and accesses to others will be blocked. After a successful state transition `PLAYER_2_CHECK` will be "unlocked" and access to `PLAYER_1_SHOOT` will be unavailable.

Reviewer 2: ```suggestion
The `PLAYER_1_SHOOT` circuit will be available first, and access to others will be blocked. After a successful state transition, `PLAYER_2_CHECK` will be "unlocked" and access to `PLAYER_1_SHOOT` will be unavailable.
```

## Compact tutorial

Compact provides the right mix of public and private data management to enable the Battleship implementation elegantly.  

### Setup

> Start with compact code setup:

Reviewer 2: compact => Compact

### Constructor

Next, set up the contructor that executes on contract deployment. For simplicity, assume that `player1` (Alice) is deploying the contract so that a few operations can be bundled together.

First validate the inputs are within the boundaries of the game:
```compact
constructor(_x1: Uint<8>, _x2: Uint<8>) {
    // input verification checks
    assert(_x1 != _x2, "Cannot use the same number twice");
    assert(_x1 > 0 && _x2 > 0, "No zero index, board starts at 1");
    assert(_x1 <= 20 && _x2 <= 20, "Out of bounds, please keep ships on the board");

}// end of constructor
```

The use of an underscore(_x1) prefixing an identifier is a good practice in cryptography to track management of private data. This does nothing for the compiler, but it helps you as a developer better track information needing to remain private. 

Then assign a Dapp specific public key for `player1`:
```compact
    // user id and assignment
    const _sk = localSk();
    const pubKey = getDappPubKey(_sk);
    player1 = disclose(pubKey);

}// end of constructor
```

> This pattern accesses a secret key from the player's private state, hashing that with a domain seperator so that user interactions can be traced inside this dapp, but not outside of it. The `getDappPublicKey` circuit will be written later.

Reviewer 2: dapp => DApp

```compact
    // state verification check
    assert(board2State == BoardState.UNSET, "There is already a player2");

}// end of acceptGame
```

This state is updated at the end of this circuit, rendering the circuit inaccessible for the remainer of the contract lifecycle.

Now verify the inputs are within bounds:
```compact
    // input verification checks
    assert(_x1 != _x2, "Cannot use the same number twice");
    assert(_x1 > 0 && _x2 > 0, "No zero index, please keep ships on the board");
    assert(_x1 <= 20 && _x2 <= 20, "Out of bounds, please keep ships on the board");

}// end of acceptGame
```

All checks have now passed, the caller is assigned as `player2`:
```compact
    // user assignment
    player2 = disclose(pubKey);

}// end of acceptGame
```

`player2` also needs to hide their private data publicly:
```compact
    // hash inputs and store them to the ledger for comparison later
    const hash1 = commitBoardSpace(_x1 as Bytes<32>, _sk);
    board2.insert(hash1);
    const hash2 = commitBoardSpace(_x2 as Bytes<32>, _sk);
    board2.insert(hash2);
    
}// end of acceptGame
```

> Now that they have committed to these values publicly, ask the the player to set their board locally:

Reviewer 2: ```suggestion
Now that they have committed to these values publicly, ask the player to set their board locally:
```

```compact
    // setting the state locally and verifying
    const localBoardState = localSetBoard(_x1, _x2);
    assert(localBoardState == BoardState.SET, 
        "Please update the state of your board to SET");

}// end of acceptGame
```

Finally, on-chain state is updated to transition to the next game state:
```compact
    // setting on-chain state
    board2State = disclose(localBoardState);
    
    // updating on-chain state
    turn = TurnState.PLAYER_1_SHOOT;

}// end of acceptGame
```

Now both players are set up with public commitments to their private data. The program will catch any attempt to present different data for ship locations in later `checkBoard` circuits.

### Hashing circuits

> Before going any further in the game flow, the necessary hashing functions can be implemented:

Reviewer 2: ```suggestion
Before going any further in the game flow, implement the necessary hashing functions:
```

Reviewer 2: Let's limit the use of passive tone in docs/tutorials.

```compact
// hashing a commitment to a board space
circuit commitBoardSpace(_x: Bytes<32>, _sk: Bytes<32>): Bytes<32> {
    const hash = persistentHash<Vector<2, Bytes<32>>>([_x, _sk]);
    return disclose(hash);
}

// hashing a Dapp specific public key to identify the user (only in this contract)
circuit getDappPubKey(_sk: Bytes<32>): Bytes<32> {
    return persistentHash<Vector<2, Bytes<32>>>([pad(32, "battleship:pk:"), _sk]);
}
```

> `commitBoardSpace` takes in a single ship location `_x` and hashes it using `persistentHash` and a `_sk`. This pattern ensures that the data in `_x` cannot be guessed. Given the small board size (20) it would relatively easy for a malicious actor to brute force hash all of the possible numbers and compare the hash with the on-chain ship locations. Hashing with complex binary data, like a `_sk`, makes the brute force attack near impossible.

Reviewer 2: ```suggestion
`commitBoardSpace` takes in a single ship location `_x` and hashes it using `persistentHash` and a `_sk`. This pattern ensures that the data in `_x` cannot be guessed. Given the small board size (20), it would be relatively easy for a malicious actor to brute-force hash all possible numbers and compare the hashes with the on-chain ship locations. Hashing with complex binary data, like a `_sk`, makes a brute-force attack nearly impossible.
```

### Shoot circuits

Back to the game operation, it is now `player1`'s turn to shoot:
```compact
export circuit player1Shoot (x: Uint<8>): [] {

    // caller verification check
    const _sk = localSk();
    const pubKey = getDappPubKey(_sk);
    assert(player1 == disclose(pubKey), "You are not player1");

}// end of player1Shoot
```
Use the same pattern of hashing the `_sk` through `getDappPubKey` to identify the user calling is the same that has already been assigned as `player1`. Any difference in the hashes indicates a different player is calling this circuit.

After verifying the caller is `player1`, verify state is as expected and inputs are valid:
```compact

    // state verification checks
    assert(board2State == BoardState.SET, "Player 2 has not yet set their board");
    assert(turn == TurnState.PLAYER_1_SHOOT, "It is not player1 turn to shoot");
    assert(winState == WinState.CONTINUE_PLAY, "A winner has already been declared");

    // input validation
    assert(x > 0 && x <= 20, "Shot out of bounds, please shoot on the board");

}// end of player1Shoot
```

:::warning

> Be aware of leaking sensitive state data through `assert` checks. The message displayed is what will be returned from the blockchain on a failed called due to a given assertion. This tutorial is explicit in these messages for learning purposes, but a malicious actor could identify sensitive data through these messages.

Reviewer 2: ```suggestion
Be aware of leaking sensitive state data through `assert` checks. The message displayed is what will be returned from the blockchain in response to a failed call due to a given assertion. This tutorial is explicit in these messages for learning purposes, but a malicious actor could identify sensitive data through these messages.
```

The final witness function needs to perform a check of a players private state data:
```typescript
    localCheckBoard: ({
        privateState
    }: WitnessContext<Ledger, BattlePrivateState>, x: bigint): [
        BattlePrivateState,
        ShotState
    ] => {
        let currentShot = ShotState.MISS;// reset to default -- MISS
        if(x == privateState.x1 || x == privateState.x2){
            currentShot = ShotState.HIT;// only HIT if it is in fact a HIT
        }
        privateState.shotState = currentShot;
        return [privateState, privateState.shotState];
    },// end of localCheckBoard
};//end of witnesses
```

It is important to note that operations on private state data in witness functions are not verified in any way. It is imperative that DApp developers always verify that computations in witness functions were performed to expectations through `assert` statements. There are rigorous checks built into the `.compact` code and the frontend test suite will verify their correct operation.

> That is everything needed for off-chain witnesses, the compiled contract can be now be exported.

Reviewer 2: ```suggestion
That is everything needed for off-chain witnesses. You can now export the compiled contract.
```

touch index.ts
```

Use the `index.ts` file to export the compiled contract and necessary types:
```typescript
import { CompiledContract } from '@midnight-ntwrk/compact-js';
import path from 'node:path';

export {
    Contract,
    ledger,
    pureCircuits,
    type Witnesses,
    type Ledger,
    type ImpureCircuits,
    type PureCircuits
} from './managed/battleship/contract/index.js';
import { Contract } from './managed/battleship/contract/index.js';
import { witnesses } from './witnesses.js';

const currentDir = path.resolve(new URL(import.meta.url).pathname, '..');
export const zkConfigPath = path.resolve(currentDir, 'managed', 'battleship');

export const CompiledBattleshipContract = CompiledContract.make(
    'BattleshipContract',
    Contract,
).pipe(
    CompiledContract.withWitnesses(witnesses),
    CompiledContract.withCompiledFileAssets(zkConfigPath),
)
```

That completes the code needed for Compact and its supporting files. To recap, the directory structure should look like this:
```
contract/
├── managed/
|   └── battleship/
|        ├── compiler/
|        ├── contract/
|        ├── keys/
|        └── zkir/
├── battleship.compact
├── index.ts
└── witnesses.ts
```

> 

Reviewer 2: Add a 'Next steps' heading and introduce the test suite tutorial for consistency with other tutorials.

> The next steps will be covered in the [Battleship Test Tutorial](./test-suite.mdx).

Reviewer 2: Anchor link text should match the destination tutorial title.
